File: probs/usaco/oldsite/dance2/dance2.cpp
#include <bits/stdc++.h>

using namespace std;

int n, m, delta;
int main()
{
	cin >> n;
	for(int iter = 0; iter < n; iter++)
	{
		delta = 0;
		cin >> m;
		string s; cin >> s;
		for(int i = 0; i < m; i++)
		{
			if(s[i] == '>') delta++;
			else delta--;
			if(delta < 0)
			{
				goto end;
			}
		}
		if(delta != 0)
		{
			goto end;
		}
		cout << "legal" << endl;
		continue;
end:
		cout << "illegal" << endl;
	}
}

File: probs/usaco/oldsite/dance2/dance2.py
n = int(input()) # number of dance lines
for _ in range(n) :
    m, s = input().split(' ') # m = length of string, s = string of cows
    m = int(m)
    delta = 0 # number of '>' cows minus number of '<' cows.
    legal = True # if the configuration is legal
    for c in s :
        if c == '>' :
            delta += 1
        else :
            delta -= 1
        if delta < 0 :
            legal = False
            break
    # Must have equal number of '>' cows and '<' cows.
    if delta != 0 :
        legal = False
    ret = 'legal'
    if not legal :
        ret = 'illegal'
    print(ret)
